libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp

```diff
@@ -0,0 +1,88 @@
+#include <utility>
+#include <stdio.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+#include "AP_TemperatureSensor_MLX90614.h"
+
+
+extern const AP_HAL::HAL &hal;
+
+
+void AP_TemperatureSensor_MLX90614::init()
+{
+    constexpr char name[] = "MLX90614";
+    (void)name;  // sometimes this is unused (e.g. HAL_GCS_ENABLED false)
+
+    if ((_params.bus_address != MLX90614_I2CDEFAULTADDR) ) {
+        printf("%s wrong I2C addr of 0x%2X, setting to 0x%2X", name, (unsigned)_params.bus_address.get(), (unsigned)MLX90614_I2CDEFAULTADDR);
+        _params.bus_address.set(MLX90614_I2CDEFAULTADDR);
+    }
+    
+    // _dev = std::move(hal.i2c_mgr->get_device(_params.bus, _params.bus_address));
+    _dev = std::move(hal.i2c_mgr->get_device(_params.bus, _params.bus_address));
+    if (!_dev) {
+        printf("MLX90614: Device is null\n");
+        return;
+        
+    }
+
+
+    WITH_SEMAPHORE(_dev->get_semaphore());
+
+    // 500 -> 2 hz
+    // 50  -> 20 hz3 
+    _dev->register_periodic_callback(50 * AP_USEC_PER_MSEC,
+                                     FUNCTOR_BIND_MEMBER(&AP_TemperatureSensor_MLX90614::_timer, void));
+    
+    printf("MLX90614: Device is ready\n");
+}
+
+
+void AP_TemperatureSensor_MLX90614::_timer()
+{
+
+    // < RAM reg - Linearized temperature, ambient
+
+     _crude_value = read_data(MLX90614_TA);
+     
+    // RAM reg - Linearized temperature, source #2 
+    // _crude_value = read_data(MLX90614_TOBJ2);
+
+    // RAM reg - Linearized temperature, source #1
+    // _crude_value = read_data(MLX90614_TOBJ1);
+    
+    
+    if(_crude_value > 0)
+    {
+        // debugg
+        // printf("MLX90614: temp = %f\n",(_crude_value * 0.02 - 273.15));
+        // temp * 0.02 - 273.15 = degrees, temp * 0.02 is temperature in kelvin
+        set_temperature((_crude_value * 0.02 - 273.15));
```

Comment:
```suggestion
        WITH_SEMAPHORE(_dev->get_semaphore());
        set_temperature((_crude_value * 0.02 - 273.15));
```

Reply:
for clarity, do this conversion computation on a separate line above it and store it into a const float

Reply:
resolved